left-sidebar: Create Collapsible Private Messages

jai2201 opened the pull request:
Fixes - #20870

jai2201 commented:
Please review and let me know about the improvements which can be made in the code written.

showell commented:
Thanks for working on this!  I am excited about this feature.

I noticed one bug that we should fix before test-deploying this on czo--if you click on "more conversations", it says "BACK TO STREAMS", which doesn't feel exactly accurate.  It's more like you are going back to the normal left sidebar view.

jai2201 commented:
> Thanks for working on this! I am excited about this feature.
> 
> I noticed one bug that we should fix before test-deploying this on czo--if you click on "more conversations", it says "BACK TO STREAMS", which doesn't feel exactly accurate. It's more like you are going back to the normal left sidebar view.

I agree @showell , I'm also rather not that much satisfied with `BACK TO STREAMS` text.
like, it can be some better, maybe saying `BACK` itself should work.

alya commented:
Mmm, yeah, I was struggling to find a good label for it, as "Back to PMs" didn't make sense either. The challenge with just "Back" is that you might have forgotten where you came from if you walked away from your Zulip window for a while, but maybe that's the best we've got for now.

jai2201 commented:
Cool, so @alya , should i replace the `BACK TO STREAMS` with `BACK` for now, or else wait for some more inputs.

timabbott commented on static/templates/pm_list_item.hbs:
```diff
@@ -13,7 +13,7 @@
             {{> status_emoji status_emoji_info}}
         </a>
-        <span class="unread_count {{#if is_zero}}zero_count{{/if}}">
+        <span id="private_message_unread_count" class="unread_count {{#if is_zero}}zero_count{{/if}}">
```
This needs to be a CSS class, not an ID -- HTML spec requires that IDs be unique in a document, and this is a thing one has more than one of.

jai2201 replied:
yes, i agree, Updated the PR.
Thanks :)

timabbott commented on static/js/click_handlers.js:
```diff
@@ -514,6 +517,36 @@ export function initialize() {
+        $("#private_messages").addClass("active_private_messages_section");
+        top_left_corner.deselect_top_left_corner_items();
+        window.location.href = "#narrow/is/private";
+    });
```
I expect we can delete this click handler in favor of just an `a href="/#narrow/is/private"`, and that'd be cleaner.

jai2201 replied:
umm, I'm not sure whether we can do it, as in this -

I've added a click handler to perform 2 functions ->
1.) Add class of active-filter to PRIVATE MESSAGES section, which just highlights it.
2.) Also at the same time i need to redirect it to `#narrow/is/private` which i cannot perform if i add `e.preventDefault()` to this click handler.

Let me know if there's an alternative in your mind for this.

timabbott commented on frontend_tests/puppeteer_tests/message-basics.ts:
```diff
@@ -259,21 +259,6 @@ async function search_tests(page: Page): Promise<void> {
-        "New stream message",
-    );
-    assert.strictEqual(await page.title(), "Private messages - Zulip Dev - Zulip");
-}
```
I don't think we should need to remove this test.

jai2201 replied:
yess, since now we've a  `ALL PRIVATE MESSAGES` icon, clicking on which takes us to `#narrow/is/private` so we can test this in `puppeteer`, 

I've added this test cases which now clicks on the icon of `all_private_messages` and expects to load All Private messages.

Thanks :))

timabbott commented on static/styles/left_sidebar.css:
```diff
@@ -157,9 +183,11 @@ li.show-more-topics {
+            color: inherit;
         }
     }
 }
```
This CSS change seems like it affects other elements; what's its motivation?

timabbott commented:
Thanks for all your work on this @jai2201!  There's a visual glitch when reloading the page that needs investigation:

![image](https://user-images.githubusercontent.com/2746074/154375320-97edfd03-4a49-4d7f-8a5d-c69c6f9e8ac6.png)

This pull request will need a lot of cleanup before it can be merged, but I think none of the issues I've noticed prevents test-deploying on chat.zulip.org at this point.

timabbott commented:
Another bug: The "more conversations" element doesn't show the count of additional unreads, which we should do like we do for topics.

![image](https://user-images.githubusercontent.com/2746074/154375401-dd2c1ce4-7e19-41f1-8a27-b7a02c01db6e.png)

jai2201 commented:
> Thanks for all your work on this @jai2201! There's a visual glitch when reloading the page that needs investigation:
> 
> ![image](https://user-images.githubusercontent.com/2746074/154375320-97edfd03-4a49-4d7f-8a5d-c69c6f9e8ac6.png)
> 
> This pull request will need a lot of cleanup before it can be merged, but I think none of the issues I've noticed prevents test-deploying on chat.zulip.org at this point.

I can't reproduce this visual glitch at all while reloading, can you please guide me on how to reproduce this basically, so that i can debug this and fix it.

jai2201 commented:
> Another bug: The "more conversations" element doesn't show the count of additional unreads, which we should do like we do for topics.
> 
> ![image](https://user-images.githubusercontent.com/2746074/154375401-dd2c1ce4-7e19-41f1-8a27-b7a02c01db6e.png)
> 
> That bug seems significant enough to be worth waiting for a fix before deploying.

Hmmm, I'm really sorry for missing this feature out of my mind, and unfortunately this got missed by `Alya` also, while testing manually!
but anyways, i've implemented this function to render `unread-count` of `more-conversations` to the DOM now!

jai2201 commented on frontend_tests/puppeteer_tests/navigation.ts:
```diff
@@ -97,7 +97,7 @@ async function navigation_tests(page: Page): Promise<void> {
     await navigate_to_subscriptions(page);
     await navigate_using_left_sidebar(page, "all_messages", "message_feed_container");
     await navigate_to_settings(page);
-    await navigate_using_left_sidebar(page, "narrow/is/private", "message_feed_container");
+    // await navigate_using_left_sidebar(page, "narrow/is/private", "message_feed_container");
     await navigate_to_subscriptions(page);
```
Also, I've this Puppeteer this back in the functionality as this also is now available.

jai2201 commented:
@timabbott , I've fixed the bug/feature you asked for adding on, to display the number of unread-counts in `more conversations` button !

jai2201 commented:
@timabbott , I request you to give this PR 1 more round of review.

timabbott commented:
You should fix the linter errors, and also add `hidden-for-spectators` to this element so the anonymous login looks right:

![image](https://user-images.githubusercontent.com/2746074/155204294-57b7842f-9ddf-4791-ab2a-a9196e45d0cc.png)

jai2201 commented:
@timabbott , Sorry for missing over the linters, 

I've fixed the linter errors.
Also, added `hidden-for-spectators` to `private_messages` div so that no PMs are visible when a Anonymous User logs in.

jai2201 commented:
@timabbott , I've pushed the changes in this PR which i mentioned in https://chat.zulip.org/#narrow/stream/101-design/topic/private.20messages.20section.20.2321102/near/1334315

I think we can test-deploy this branch once again now and have more reviews.

jai2201 commented on static/js/resize.js:
```diff
@@ -203,7 +203,10 @@ export function resize_bottom_whitespace(h) {
+    let height = h.stream_filters_max_height;
+    height = height - $("#private_messages").safeOuterHeight(true);
+    $("#stream-filters-container").css("max-height", height);
+    $("#private-container").css("max-height", h.stream_filters_max_height - 50);
```
There's 1 thing in latest commit i would like to mention right now -> 

The issue reported for "Scrolling seems to be broken in `more conversations` view" was caused because i missed to allot a `max-height` CSS property to PM section, and hence the PMs went with `overflow:hidden`;

So now to fix it, we would need to allot a max-height here, for which i've used `50px` right now; which we might have to change/use a proper algorithm to calculate it, if we face any reported issues related to scrolling in more PMs section,

But for now it seemed to work properly in my local dev which is why i'm asking for a test-deploy of it; but yes, i won't be surprised with a reported bug on this.

jai2201 commented:
Also, we're still not sure on how to arrange the commits of this branch for merging in main; but it doesn't stop us from test-deploying so i haven't focused much right now on arranging the commits properly.

timabbott commented:
@jai2201 can you start working on the node tests for this?  It's annoying to keep this deployed on chat.zulip.org with test coverage failing.

alya commented:
1. In the expanded PMs view, "Private messages" should be bold when the user is in the "all PMs".
2. When the PMs section is expanded, the tooltip on the triangle to collapse it should be "Collapse private messages", not "Expand private messages".

Otherwise looks good to me!

jai2201 commented:
> 1. In the expanded PMs view, "Private messages" should be bold when the user is in the "all PMs".
> 2. When the PMs section is expanded, the tooltip on the triangle to collapse it should be "Collapse private messages", not "Expand private messages".
> 
> Otherwise looks good to me!

Implemented both of these features.

Thanks :)

jai2201 commented:
This PR would need cleanup on the commits; 
as after taking on the localstorage thing into implementation, the earlier commits now need different explanations.

timabbott commented on static/js/click_handlers.js:
```diff
@@ -532,6 +535,39 @@ export function initialize() {
+            // We resize the STREAMS section after changing the state of PM section.
+            resize.resize_stream_filters_container();
+        }, 0);
+    });
```
This is a more complex click handler than we usually aim to have in this file. This click handler should call a function with a name like `pm_list.toggle_private_messages_section`.

jai2201 replied:
Sure, would update this in a preparatory commit.

timabbott commented on static/js/click_handlers.js:
```diff
@@ -532,6 +535,39 @@ export function initialize() {
+            // To expand the PM section list we need the PM section state to be collapsed
+            // then there comes 2 possible conditions regarding the number of PMs present in the section
+            // either it should 0 PMs present or else only 1 which is active PM can be present outside the collapsed PM section.
+            pm_list.expand();
```
I don't understand this comment; does it belong before the `if` statement?

jai2201 replied:
Hmm, this was a confusing comment, fixed it in #21726.

timabbott commented on static/js/compose.js:
```diff
@@ -250,6 +251,9 @@ export function send_message(request = create_message_object()) {
     transmit.send_message(request, success, error);
+    if (request.type === "private" && $(".expanded-private-messages").children().length === 0) {
+        pm_list.append_new_active_pm_in_collapsed_pms();
+    }
```
This is the wrong place to add this, because messages can be sent on other clients. This belongs in the `server_events_dispatch.js` code path, likely in whatever "update the pm_list" function is called when we receive a new message from the server (including ones we sent).

You can test for this sort of thing easily by using two browsers in your development environment to send messages.

BTW the new application feature tutorial is worth reading, as it explains this category of bug and why it's important to avoid it (especially https://zulip.readthedocs.io/en/latest/tutorials/new-feature-tutorial.html#update-the-frontend)

jai2201 replied:
Thanks Tim, Yes, this was not needed in code, there was some other bug which i handled with #21726.

timabbott commented on static/js/narrow.js:
```diff
@@ -833,6 +835,7 @@ function handle_post_narrow_deactivate_processes() {
     }
 
     top_left_corner.handle_narrow_deactivated();
+    pm_list.handle_narrow_deactivated();
```
The change to extract a `handle_narrow_deactivated` function (moving the `pm_list` related code currently in `top_left_corner.handle_narrow_deactivated`) without any other changes would be a good prep commit that we can split out and integrate before the rest of this PR to reduce its complexity.  There may be a few others as well.

jai2201 replied:
Got it, Sure -- this can come out earlier as a prep commit to reduce this PR's complexity.

timabbott commented on static/js/pm_list.js:
```diff
@@ -108,19 +142,54 @@ export function _get_convos() {
+    return {
+        key,
+        more_items: true,
+        more_convos_unread_count,
```
As in https://github.com/zulip/zulip/pull/21357, we'll want to replace all references to "convos" with "conversations".

jai2201 replied:
Sure.

timabbott commented on static/js/pm_list.js:
```diff
@@ -147,15 +225,149 @@ export function is_all_privates() {
+
+export function activate_all_private_messages_view() {
+    $("#private_messages_section").addClass("active_private_messages_section");
+    $(".more-private-messages-sidebar-title").css("font-weight", "bold");
```
Don't set CSS this way -- set a semantic CSS class, which can happen to use bold as its currently styling.

jai2201 replied:
OK, would change this :)

timabbott commented on static/js/pm_list.js:
```diff
@@ -147,15 +225,149 @@ export function is_all_privates() {
+    update_private_messages();
+    setTimeout(() => {
+        resize.resize_stream_filters_container();
+    }, 0);
```
This type of `setTimeout` is a code smell and often represents a bug; please document with a comment why the delay is necessary.

jai2201 replied:
I understand that this type of code is a mess and was ofc difficult for me also to debug to this solution  -- I've seen this sort of thing implemented at some other places as well, In any case I'll comment out a reason behind this.

timabbott commented on static/js/pm_list_dom.js:
```diff
@@ -19,13 +19,13 @@ export function keyed_pm_li(convo) {
+export function pm_ul(nodes) {
     const attrs = [
-        ["class", "expanded_private_messages"],
+        ["class", "expanded-private-messages"],
```
Why the rename from `expanded_private_messages` to `expanded-private-messages`?  In any case, I think the variable/class renames here would all be great preparatory commits that can be merged independently of the feature to shrink this diff.

jai2201 replied:
Hmm -- this also can surely come out as a prep commit, thanks :)

timabbott commented on static/js/recent_topics_ui.js:
```diff
@@ -649,6 +650,7 @@ export function show() {
     narrow_state.reset_current_filter();
     narrow.set_narrow_title("Recent topics");
     message_view_header.render_title_area();
+    pm_list.handle_narrow_deactivated();
```
This should go next to `stream_list.handle_narrow_deactivated`, above, unless there's a very good reason for it not to.

jai2201 replied:
Added a comment to answer the reason in #21726

timabbott commented on static/js/stream_list.js:
```diff
@@ -437,6 +438,7 @@ export function update_stream_sidebar_for_narrow(filter) {
 
     if (!info.topic_selected) {
         stream_li.addClass("active-filter");
+        pm_list.handle_narrow_deactivated();
```
This seems like a very strange place to add this, given what other functions we call nearby -- is this necessary?  It seems most likely that the bug is that we're doing this in the wrong place.

jai2201 replied:
Hmm, Thanks for this Tim, fixed this bug too.

timabbott commented on static/js/top_left_corner.js:
```diff
@@ -79,34 +56,27 @@ export function handle_narrow_activated(filter) {
         if (filter_name === "starred") {
             filter_li = $(".top_left_starred_messages");
             filter_li.addClass("active-filter");
+            pm_list.handle_narrow_deactivated();
```
These calls also seem very likely to be working around some other bug in your code -- we should not be needing to call a "handle_narrow_deactivated" function inside another `handle_narrow_activated` handler's code path.

jai2201 replied:
Hmm, fixed it by removing these and handle the other bug in #21726.

timabbott commented on static/styles/left_sidebar.css:
```diff
@@ -62,6 +62,52 @@ li.show-more-topics {
     }
 }
 
+#show_more_pms {
```
I think this should probably have a name more like `li.show-more-private-conversations`, for similarity to `show-more-topics`.

timabbott replied:
Also, could we share CSS with `show-more-topics`, rather than copying it?

jai2201 replied:
Hmm -- would fix this.

timabbott commented on static/styles/left_sidebar.css:
```diff
@@ -62,6 +62,52 @@ li.show-more-topics {
+    width: 90%;
+}
+
+#toggle_private_messages_section_icon {
```
I think we want this CSS to be for a  `.toggle_sidebar_section_icon` class, since I imagine we're going to end up wanting to reuse this toggling feature for other sections.

timabbott commented on static/styles/left_sidebar.css:
```diff
@@ -478,16 +508,6 @@ li.expanded_private_message {
-    .fa-chevron-left {
-        text-decoration: none;
-    }
-}
```
I think you should extract any commits that intend to refactor the existing CSS for "more topics" into independent, correct preparatory commits; it's very hard to figure out where this code ends up.

jai2201 replied:
Done in #21726

timabbott commented on static/templates/left_sidebar.hbs:
```diff
@@ -69,6 +55,39 @@
             </li>
         </ul>
+
+        <div id="private_messages" class="zoom-out hidden-for-spectators">
```
`sidebar-private-messages-container` is a more reasonable name for this.

jai2201 replied:
OK, would go further with this name.

timabbott commented:
@jai2201 thanks for your work on this. I've done a review pass through this, marking things I could easily identify as problems or naming convention violations. Once we merge #21357, which I'm hoping will be shortly, I think your next goal should be to extract a clean PR that does a series of refactors to the existing implementation that we can merge with no functional side effects to decrease the complexity of this PR; I've made several suggestions above. I don't intend to re-review whether you've correctly implemented changes like the `server_events` piece before we try to integrate those preparatory refactoring PRs, but I encourage you to update this to a working version.

The other thing I should say is that I don't think there's any reason not to squash the changes you have here into a single commit as a starting point. It'll make rebasing easier, and fundamentally the commits you have are implemented in a workflow of "Each commit fixes bugs in the previous commits", which isn't useful commit history. As part of working towards merging this, we'll end up wanting to split out chunks like #21357 and the suggestions I've made here that we can merge without breaking the project (please test and reason carefully that each individual commit works fully).

This PR is going to likely need a lot of rounds of work before it'll be in a state that we can consider merging; in particular there's a number of things I flagged above that suggest there's a set of live-update bugs that we haven't understood and have instead papered over.

The other thing that I think we'll want to split is that it appears a substantial portion of the complexity remaining in this pull request is related to being able to show one conversation even when "private messages" is collapsed; perhaps we should cut that feature for the moment and just have "If you narrow to a private message conversation, private messages is automatically uncollapsed" as an intermediate checkpoint?  I think doing so would probably hasten our ability to integrate the core feature here by weeks.

jai2201 commented:
@timabbott , thanks a lot for the detailed feedback on this PR -- I've tried to answer most of your comments/suggestions above (was not able to answer them all as i've lost a bit touch with the reasons behind those lines, would update them as soon as i get back to work on this PR i.e. this week itself)

And coming to the comment of --> Dropping out the feature of active PM being present out of the collapsed PM section, I think we can go forward with including that -- as we're already done with the test cases for that feature, but still we can make a decision over it once I have few of the preparatory commits merged over this PR.

jai2201 commented:
I've raised a PR #21726 which has all these comments answered and with proper arrangements of commits.

alya commented:
We should consider adding the first point of #22113 as a release goal once this is merged, to make the feature feel more complete.

timabbott commented:
Closing as this was completed in https://github.com/zulip/zulip/pull/21726.
